### PR TITLE
Enhancement, add the field type conflict check in semantic check

### PR DESCRIPTION
**Pull request by penghuo**
*Issue #, if available:* #348, #445 

*Description of changes:*
* Add wildcard index handing in semantic. If the field is used in query have conflict type in wildcard index, semantic exception will be thrown.
* The mapping conflict resolver logic in the TermFieldRewriter has been refined to do the lazy evaluation only when the field is accessed.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/antlr/semantic/scope/TypeSupplier.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    /**
+     * Get the {@link Type}
+     * Throw {@link SemanticAnalysisException} if conflict found.
+     * Currently, if the two type not equal, it been treated as conflict.
+     */
+    @Override
+    public Type get() {
+        if (types.size() > 1) {
+            throw new SemanticAnalysisException(
+                    String.format("Symbol [%s] have conflict type [%s]", symbolName, types));
```
Comment (abbashus) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/antlr/semantic/scope/TypeSupplier.java:56]: We should change this to `Field [%s] has conflicting type [%s]`

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/antlr/semantic/scope/TypeSupplier.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+    public TypeSupplier add(Type type) {
+        types.add(type);
+        return this;
+    }
+
+    /**
+     * Get the {@link Type}
+     * Throw {@link SemanticAnalysisException} if conflict found.
+     * Currently, if the two type not equal, it been treated as conflict.
```
Comment (abbashus) [inline src/main/java/com/amazon/opendistroforelasticsearch/sql/antlr/semantic/scope/TypeSupplier.java:50]: np: `type` - >`types` , `it been treated as conflict` - > `they are treated as conflicting`

File: src/test/java/com/amazon/opendistroforelasticsearch/sql/antlr/semantic/SemanticAnalyzerFieldTypeTest.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    public void mixNonConflictTypeAndConflictFieldTypeShouldFailSemanticCheck() {
+        expectValidationFailWithErrorMessages("SELECT id, age FROM account* WHERE id = 1",
+                "Symbol [age] have conflict type");
+    }
+
+    /**
+     * age has different type in account1 and account2.
+     */
+    @Test
+    public void ConflictFieldTypeWithAliasShouldFailSemanticCheck() {
```
Comment (abbashus) [inline src/test/java/com/amazon/opendistroforelasticsearch/sql/antlr/semantic/SemanticAnalyzerFieldTypeTest.java:67]: np: `Conflict` -> `conflict`

File: src/test/java/com/amazon/opendistroforelasticsearch/sql/antlr/semantic/scope/TypeSupplierTest.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+package com.amazon.opendistroforelasticsearch.sql.antlr.semantic.scope;
+
+import com.amazon.opendistroforelasticsearch.sql.antlr.semantic.SemanticAnalysisException;
+import com.amazon.opendistroforelasticsearch.sql.antlr.semantic.types.base.ESDataType;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.ExpectedException;
+
+import static org.junit.Assert.*;
```
[the remark this answers is not in the document]
Comment (penghuo) [reply src/test/java/com/amazon/opendistroforelasticsearch/sql/antlr/semantic/scope/TypeSupplierTest.java:24]: done.

File: src/test/java/com/amazon/opendistroforelasticsearch/sql/antlr/semantic/SemanticAnalyzerFieldTypeTest.java (hunk 3 above)
Comment (penghuo) [reply src/test/java/com/amazon/opendistroforelasticsearch/sql/antlr/semantic/SemanticAnalyzerFieldTypeTest.java:67]: done.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/antlr/semantic/scope/TypeSupplier.java (hunk 2 above)
Comment (penghuo) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/antlr/semantic/scope/TypeSupplier.java:50]: done.

File: src/main/java/com/amazon/opendistroforelasticsearch/sql/antlr/semantic/scope/TypeSupplier.java (hunk 1 above)
Comment (penghuo) [reply src/main/java/com/amazon/opendistroforelasticsearch/sql/antlr/semantic/scope/TypeSupplier.java:56]: done.

**Pull request merged**